### PR TITLE
Have statustext messages sent to the Frsky lib queue

### PR DESCRIPTION
**Pull request by floaledm**
Suggested replacement of #4864.
Followed examples from the use of DataFlash in GCS_MAVLINK and of airspeed in AP_AHRS.
Tested working.

### CODE REVIEW

**Comment by floaledm** [comment]
Thanks for your review. We discussed in #4864, and I agree, that there's a number of libs that should be selected during compilation, but until some rewrite of the build system, this seems like the next best thing because it's a method that's already used: in AP_AHRS, the presence of AP_Airspeed is checked at runtime. AP_Airspeed is included in the builds of all vehicule types but it's only used in Plane...

**Comment by tridge** [comment]
I think it would be better if the init() function in the frsky lib setup the callback in the mavlink lib. Otherwise it will be doing the work to queue the data in the frsky lib even if no uart has been configured with the frsky protocol.

File: Tools/Replay/Replay.h [hunk 1]
```diff
@@ -51,7 +51,6 @@
 #include <signal.h>
 #include <unistd.h>
 #include <AP_HAL/utility/getopt_cpp.h>
-#include <AP_SerialManager/AP_SerialManager.h>
```
Comment (floaledm) [inline Tools/Replay/Replay.h]: This is just a quick cleanup because it is included twice...

**Comment by floaledm** [comment]
@tridge Thanks for taking thhe time to review the code. I implemented your excellent suggestion. Let me know if any other change is necessary.

**Comment by OXINARF** [comment]
Has this been compiled with waf with the FrSky telemetry enabled? I doubt that.

**Comment by tridge** [comment]
I'd actually prefer to remove all the ifdefs and always compile this in. That would be a separate PR though and we'd need to check with Randy if he is OK with that.

**Comment by floaledm** [comment]
Let me know once you check with Randy, and I'd be happy to put together that other PR.

**Comment by OXINARF** [comment]
@tridge I think that would be fine once #4744 is fixed.

**Comment by floaledm** [comment]
@OXINARF Why would #4744 need to be fixed before removing all the `FRSKY_TELEM_ENABLED` statements?

**Comment by OXINARF** [comment]
@floaledm Because a bug exists and removing FRSKY_TELEM_ENABLED removes the possibility of removing the library from the build.

**Comment by floaledm** [comment]
I would like to know what is causing this and fix it so that there's no potential inconvenience for anyone,  but I also think it's important to keep in mind that:
- #4744 was only reported by lvale (our beta testers using PixRacer haven't had the issue and we haven't been able to replicate based on the info lvale provided)
- would only potentially impact actual users of FrSky passthrough protocol (those having a telemetry cable installed and ArduPilot configured to use passthrough)
- would only happen with OpenTX versions prior to 2.2 and while in discovery (even on 2.1, discovery is not necessary to make the system work)
- would not be avoided or solved by holing off on this or other PRs we've been discussing
- may not even occur for lvale anymore (this was reported on 8/29/16 master)

So, I'll ask lvale again to test again and hopefully he has time to perform some troubleshooting, but I don't see a good reason to put all FrSky lib related PRs on hold in the meantime...

**Comment by magicrub** [comment]
He did not mean to halt PRs, Fernando was responding to the content about
the removal of the define for copter. This issue does not effect other PRs.

**Comment by magicrub** [comment]
Oh my mistake, I got PRs mixed up. Disregard my previous comment.

**Comment by OXINARF** [comment]
>  I don't see a good reason to put all FrSky lib related PRs on hold in the meantime...

Where have I said that?

**Comment by floaledm** [comment]
That's good news that I misunderstood. If you don't have any issues with this PR, let's move the discussion on `FRSKY_TELEM_ENABLED` to #4744 ...

**Comment by OXINARF** [comment]
I still don't like this, it is a hack. The FrSky library is duplicating the messages from the MAVLink library. The Display PR also wants to do it. Then someone comes with another telemetry option and wants to do it too and it will just keep growing.

**Comment by floaledm** [comment]
I agree with you that a rewrite/restructuring of GCS_MAVLink and AP_Frsky_Telem (and possibly others) into a generic telemetry lib would be in order (at some point in time), but few would have the experience to do that, so maybe that would be a good proposal subject? http://discuss.ardupilot.org/c/proposals

**Comment by floaledm** [comment]
I should add that having a generic telemetry lib in and of itself wouldn't address your concern, which is about passing statustext messages to other parts of the code (parts not related to sending statustext messages via MAVLink):
- Statustext messages get stored in a ringbuffer and the messages get removed once they are transmitted on a given telemetry link. I think you would still need to duplicate the messages in the different ringbuffers (one per telemetry link) to keep track of what has been sent for each link, given how the ringbuffer is setup.
- Statustext messages are passed to the DataFlash lib, and I don't know what would be a way around that.

**Comment by tridge** [comment]
I agree we need some better structure around text messages to users, but I think we should put this in now, and migrate to the new structure once we have it.
I'll push this now

**Comment by OXINARF** [comment]
> I agree with you that a rewrite/restructuring of GCS_MAVLink and AP_Frsky_Telem (and possibly others) into a generic telemetry lib would be in order (at some point in time), but few would have the experience to do that

Yes, I agree that you need some experience to do it. But I think that the more hacks go in, the hardest and unlikely to get done it gets.

> I should add that having a generic telemetry lib in and of itself wouldn't address your concern, which is about passing statustext messages to other parts of the code

That's obviously part of the work for having a telemetry library. Messages shouldn't be passed around, but instead values and flags. At best you would have a vehicle function that translates a value/flag to a message.

**Pull request merged**